src/main/ruby/core/enumerable.rb

```diff
@@ -716,8 +716,23 @@ def max_n(n, &block)
   end
   private :max_n
 
-  def max_by
+  def max_by_n(n, &block)
+    n = Rubinius::Type.rb_num2long(n)
+    raise ArgumentError, "negative size #{n}" if n < 0
+    return [] if n == 0
+    pairs = []
+    each do
+      object = Truffle.single_block_arg
+      result = block.call(object)
+      pairs << [result, object]
+    end
+    pairs.sort_by {|p| p[0]}.reverse.first(n).map {|p| p[1]}
```

Comment:
This implementation seems compatible, but it's quite inefficient for a small `n`.
What do MRI/JRuby do here?

Reply:
@eregon Fixed this to avoid creating the `pairs` array.